### ISSUE TITLE
ubuntu java package has broken cacerts

### ISSUE THREAD

**Issue by niloc132** [docker-library/java#19]
See https://bugs.launchpad.net/ubuntu/+source/ca-certificates-java/+bug/1396760

At this time, the latest image id aec8201c9d63 has no /etc/ssl/certs/java/cacerts file:

    $ docker run --rm -it java:openjdk-8-jdk ls -al /etc/ssl/certs/java/cacerts
    ...
    ls: cannot access /etc/ssl/certs/java/cacerts: No such file or directory

Note that this is not broken in java 7
    $ docker run --rm -it java:openjdk-7-jdk ls -al /etc/ssl/certs/java/cacerts
    ...
    -rw-r--r-- 1 root root 206373 Jan  1 07:13 /etc/ssl/certs/java/cacerts

This means that https cannot be used in these images (among other issues), which breaks downstream images like maven:3.2-jdk-8 (which is where I discovered the issue).

A workaround (from the linked issue) is to add command to the build, after the apt-get command:

    /var/lib/dpkg/info/ca-certificates-java.postinst configure

Note that `sudo` is left off, as the image doesn't have sudo, and runs as root anyway.

**Comment by yosifkit** [comment]
@tianon should we implement the workaround to fix jdk 8 images and any children?

**Comment by tianon** [comment]
I suppose, although only if it's going to be temporary for sure.  Has anyone figured out what's wrong upstream to cause it?  I'd feel a lot better about carrying a temporary workaround if we made sure we did our due diligence and got a patch submitted upstream so we make sure it really is only temporary.

**Comment by md5** [comment]
Looks like `/etc/ca-certificates/update.d/jks-keystore` can be used to generate `/etc/ssl/certs/java/cacerts`. I haven't dug in, but I'm assuming it generates it from the certs installed by `ca-certificates`. All of this is part of the `ca-certificates-java` package.

**Comment by md5** [comment]
It looks like the issue is that there is no `java` in the `$PATH`. There are errors during the installation of both `ca-certificates` and `ca-certificates-java` that get ignored. Right after `ca-certificates` gets set up, *then* `update-alternatives` gets called to make `java` work.

```
Setting up ca-certificates-java (20140324) ...
/var/lib/dpkg/info/ca-certificates-java.postinst: line 53: java: command not found
/var/lib/dpkg/info/ca-certificates-java.postinst: line 66: java: command not found
done.
...
Processing triggers for ca-certificates (20141019) ...
Updating certificates in /etc/ssl/certs... 173 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d....
/etc/ca-certificates/update.d/jks-keystore: 82: /etc/ca-certificates/update.d/jks-keystore: java: not found
E: /etc/ca-certificates/update.d/jks-keystore exited with code 1.
done.
Setting up openjdk-8-jre-headless:amd64 (8u40~b09-1) ...
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/rmid to provide /usr/bin/rmid (rmid) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java to provide /usr/bin/java (java) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/keytool to provide /usr/bin/keytool (keytool) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/jjs to provide /usr/bin/jjs (jjs) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/pack200 to provide /usr/bin/pack200 (pack200) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/rmiregistry to provide /usr/bin/rmiregistry (rmiregistry) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/unpack200 to provide /usr/bin/unpack200 (unpack200) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/orbd to provide /usr/bin/orbd (orbd) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/servertool to provide /usr/bin/servertool (servertool) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/tnameserv to provide /usr/bin/tnameserv (tnameserv) in auto mode
update-alternatives: using /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jexec to provide /usr/bin/jexec (jexec) in auto mode
```

**Comment by md5** [comment]
I think I found the root cause here: http://anonscm.debian.org/viewvc/pkg-java/trunk/ca-certificates-java/debian/postinst.in?view=markup#l29

It looks like the postinstall script attempts to find its `java` executable by shortcutting `update-alternatives` and looking for the JDK directly. To do this, it has a hard-coded list of possible JDK locations that doesn't include JDK 8, that `java` command will not be used.

After the installation is complete (in particular the selection of JDK 8 as `/usr/bin/java` in its postinstall), the `update-ca-certificates` script will now work correctly because it can find `java` from `$PATH`.

It looks like one fix would be to update the postinstall script to look for the JDK in `java-8-openjdk-$arch` and `java-8-openjdk`. That will just break for the next JDK, but it fixes what's being done currently at least.

**Comment by yosifkit** [comment]
Proposal: Add the workaround:

```Dockerfile
RUN /var/lib/dpkg/info/ca-certificates-java.postinst configure
```

But also add `ca-certificates-java` to the `apt-get install` with the specific failing version `20140324`, that way we can update the version and drop the `RUN` when it gets fixed (or just drop both).  Either way we get a nice docker cache bust.

**Comment by tianon** [comment]
For the record, @yosifkit filed a patch upstream to update the lines @md5 pointed out: https://bugs.debian.org/775775 :heart:

I'm +1 on @yosifkit's proposal for now, especially since we've got a nice natural cache-bust when it's fixed upstream. :+1:

**Comment by bilalakil** [comment]
Hello folks, I just wanted to point out that this workaround may not be necessary anymore, and thus the code for it may be able to be removed from the Dockerfile.

Here's my case:

```
FROM debian:jessie
MAINTAINER Bilal Akil <mail@bilalakil.me>
 
# Mostly copied from `docker-library/openjdk`'s
# `openjdk-8-jdk` Dockerfile, which uses Debian!
#
# Referenced during development:
#   https://github.com/docker-library/openjdk/blob/2dd0496901a9c01c78895d0e6618b36c08f
78bde/openjdk-8-jdk/Dockerfile
# Latest:
#   https://github.com/docker-library/openjdk/blob/master/openjdk-8-jdk/Dockerfile
 
ENV DBUILD_OPENJDK_VERSION 8u72-b15-1~bpo8+1
ENV DBUILD_CA_CERTIFICATES_JAVA_VERSION 20140324
 
ENV JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64
ENV JAVA_VERSION 8u72
 
RUN \
  # We need to add OpenJDK's backport repository to find their Java 8.
  echo 'deb http://httpredir.debian.org/debian jessie-backports main' > /etc/apt/sourc
es.list.d/jessie-backports.list \
  \
  # Now we can `apt-get` it.
  && apt-get update \
  && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
    openjdk-8-jdk="$DBUILD_OPENJDK_VERSION" \
    ca-certificates-java="$DBUILD_CA_CERTIFICATES_JAVA_VERSION" \
  \
  # See https://github.com/docker-library/java/issues/19#issuecomment-70546872
  # (This has been commented out for demonstration.)
  # /var/lib/dpkg/info/ca-certificates-java.postinst configure
  \
  # And finally we can reverse as much of the damage as we can.
  && rm /etc/apt/sources.list.d/jessie-backports.list \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

This is my Dockerfile. It was originally using a different base image but for the purpose of this example `debian:jessie` is fine. If I understand correctly, that's the same base operating system as in the OpenJDK 8 Dockerfile.

```
$ docker build -t test .
$ docker run --rm test ls -la /etc/ssl/certs/java/cacerts
-rw-r--r-- 1 root root 208916 Feb 11 07:32 /etc/ssl/certs/java/cacerts
```

So yeah, it doesn't seem necessary anymore..?

**Comment by tianon** [comment]
Hmm, that's strange -- the Debian bug is still open, and was set to severity "serious" just days ago by one of the Java maintainers. :confused:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=775775#13

> **Severity set to 'serious' from 'normal'** `Request was from Matthias Klose <doko@debian.org> to control@bugs.debian.org. (Mon, 08 Feb 2016 20:51:06 GMT)`

**Comment by hayday100** [comment]
I am trying to use the workaround in Ubuntu 14.04 since it obviously has the problem.  I put it in the dockerfile.
RUN /var/lib/dpkg/info/ca-certificates-java.postinst configure
before
RUN apt-get install -y default-jdk
but I got
/bin/sh: 1: /var/lib/dpkg/info/ca-certificates-java.postinst: not found

The command works from terminal directly but that certainly does not help docker to build.  Any sample dockerfile?

**Comment by niloc132** [comment]
@hayday100 from my original post, you must put

    RUN /var/lib/dpkg/info/ca-certificates-java.postinst configure

_after_ the apt-get install command. If you do it before, then the file won't yet have been created, since it is created by installing java.

**Comment by hayday100** [comment]
I tried that last night too.  But it certainly got stuck at the 

`Setting up ca-certificates-java (20130815ubuntu1) ...`

which obviously beat the purpose of the workaround.

**Comment by md5** [comment]
@hayday100 That's not the purpose of this workaround. This workaround existed because `cacerts` was not being created. I actually think the fact that you're getting that hanging behavior shows that the package trigger that creates `cacerts` is trying to run.

The problem you appear to be running into is https://github.com/docker/docker/issues/18180

Could you provide the output of `docker version`?

**Comment by hayday100** [comment]
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

**Comment by hayday100** [comment]
It is probably irrelevant but it is a VirtualBox 5.0.14 in Windows 8.1.

**Comment by hayday100** [comment]
I am willing to upgrade Ubuntu if that helps but people say that this persists until Ubuntu 15.04 so that's why I am very interested in this workaround.  But it looks like the workaround does not help in docker since there is not a right place for it in the dockerfile.

**Comment by md5** [comment]
It's odd that it's showing Linux for both server and client in that case.
Did you SSH into the VM to run “docker version”?

Regardless, I thought that AUFS-related bug in Boot2docker was fixed with
the 1.10 versions... Maybe it's something else, but I believe it's still
unrelated to the workaround being discussed.

**Comment by hayday100** [comment]
No.  It is a local VirtualBox that I installed from the ISO from scratch.

**Comment by md5** [comment]
Actually, are you running Ubuntu on your VM as opposed to Boot2docker?

**Comment by hayday100** [comment]
I am creating the VM for this in case some one is interested in reproducing the error.
https://www.ibm.com/developerworks/cloud/library/cl-bluemix-minecraft-docker-trs-1/

**Comment by hayday100** [comment]
Do you think that boot2docker can be an alternative for Ubuntu in light of this ca-certificates problem?

**Comment by hayday100** [comment]
The lack of GUI in Boot2docker will make it difficult for part 2 of the same tutorial but possible.  I just don't want to chase a dead end.

**Comment by hayday100** [comment]
I am a docker rookie.  According to my understanding, he docker server is where the images are and client is where the containers will run from the images generated.  The error occurs when the docker server is trying to add a ssh certificate into the image yet the openjdk certificate is not recognized. I assume if docker is a full OS, the workaround will definitely work.  But it is a mini-VM where things are built only from dockerfile following a sequential order (or at least before I know enough to tweak it).

**Comment by bilalakil** [comment]
Hi @hayday100, I've experienced a very similar problem as you (where running a Docker container hangs on `Setting up ca-certificates-java...` and think @md5 is correct about the issue being [docker/docker/issues/18180](https://github.com/docker/docker/issues/18180).

It's a tricky problem and took me quite a while to deal with, however workarounds were covered extensively in that issue - look for @AkihiroSuda's big "LATEST QUICK WORKAROUNDS" comment (NB: is there any way for me to copy a link to the comment? For now just search the previously quoted words). Note that there's more than one of said comment from this generous contributor, so ensure you're reading the most up to date one.

In summary there's a kernel bug that has reached many operating systems and affects particular versions of Docker. That is, it's the kernel of the OS _running Docker_, not the OS that _Docker is running_ that has the problem. The workaround is to either up/downgrade your kernel to one without that problem, or downgrade Docker instead. I wasn't feeling up to playing with the kernel so I downgraded my Docker to 1.8.3 and this problem disappeared. Obviously that's just a short-term solution, but hopefully that'll help you deal with your problem.

**Comment by hayday100** [comment]
It is well said.  I downgraded the kernel and the hanging is gone.  But I am facing another error.  Would any one kindly point my way again?  Thanks to all!
`Sending build context to Docker daemon 8.326 MB
Step 1 : FROM ubuntu:14.04
 ---> 14b59d36bae0
Step 2 : MAINTAINER Kyle Brown “brownkyl@us.ibm.com”
 ---> Using cache
 ---> 3ef5cf701bbf
Step 3 : RUN apt-get update
 ---> Using cache
 ---> 4e199f877890
Step 4 : RUN apt-get install -y wget
 ---> Running in 9902f96e33f6
Reading package lists...
Building dependency tree...
Reading state information...
The following extra packages will be installed:
  ca-certificates libidn11 openssl
The following NEW packages will be installed:
  ca-certificates libidn11 openssl wget
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 1040 kB of archives.
After this operation, 2419 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu/ trusty/main libidn11 amd64 1.28-1ubuntu2 [93.0 kB]
Get:2 http://archive.ubuntu.com/ubuntu/ trusty-updates/main openssl amd64 1.0.1f-1ubuntu2.16 [488 kB]
Get:3 http://archive.ubuntu.com/ubuntu/ trusty-updates/main ca-certificates all 20141019ubuntu0.14.04.1 [189 kB]
Get:4 http://archive.ubuntu.com/ubuntu/ trusty-updates/main wget amd64 1.15-1ubuntu1.14.04.1 [269 kB]
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Fetched 1040 kB in 0s (1893 kB/s)
Selecting previously unselected package libidn11:amd64.
(Reading database ... 11542 files and directories currently installed.)
Preparing to unpack .../libidn11_1.28-1ubuntu2_amd64.deb ...
Unpacking libidn11:amd64 (1.28-1ubuntu2) ...
Selecting previously unselected package openssl.
Preparing to unpack .../openssl_1.0.1f-1ubuntu2.16_amd64.deb ...
Unpacking openssl (1.0.1f-1ubuntu2.16) ...
Selecting previously unselected package ca-certificates.
Preparing to unpack .../ca-certificates_20141019ubuntu0.14.04.1_all.deb ...
Unpacking ca-certificates (20141019ubuntu0.14.04.1) ...
Selecting previously unselected package wget.
Preparing to unpack .../wget_1.15-1ubuntu1.14.04.1_amd64.deb ...
Unpacking wget (1.15-1ubuntu1.14.04.1) ...
Setting up libidn11:amd64 (1.28-1ubuntu2) ...
Setting up openssl (1.0.1f-1ubuntu2.16) ...
Setting up ca-certificates (20141019ubuntu0.14.04.1) ...
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
Setting up wget (1.15-1ubuntu1.14.04.1) ...
Processing triggers for libc-bin (2.19-0ubuntu6.7) ...
Processing triggers for ca-certificates (20141019ubuntu0.14.04.1) ...
Updating certificates in /etc/ssl/certs... 173 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d....done.
 ---> e4524d12714e
Removing intermediate container 9902f96e33f6
Step 5 : RUN mkdir minecraft
 ---> Running in 439745a60af1
 ---> c374b318cdeb
Removing intermediate container 439745a60af1
Step 6 : RUN wget -O minecraft/minecraft_server.jar https://s3.amazonaws.com/Minecraft.Download/versions/1.8.9/minecraft_server.1.8.9.jar
 ---> Running in 3e88b3ccc3f3
--2016-02-22 19:23:17--  https://s3.amazonaws.com/Minecraft.Download/versions/1.8.9/minecraft_server.1.8.9.jar
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.13.232
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.13.232|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8320755 (7.9M) [application/octet-stream]
Saving to: 'minecraft/minecraft_server.jar'

     0K .......... .......... .......... .......... ..........  0%  875K 9s
    50K .......... .......... .......... .......... ..........  1% 1.01M 8s
   100K .......... .......... .......... .......... ..........  1% 1.64M 7s
   150K .......... .......... .......... .......... ..........  2% 2.48M 6s
   200K .......... .......... .......... .......... ..........  3% 3.06M 5s
   250K .......... .......... .......... .......... ..........  3% 2.02M 5s
   300K .......... .......... .......... .......... ..........  4% 8.00M 4s
   350K .......... .......... .......... .......... ..........  4% 2.35M 4s
   400K .......... .......... .......... .......... ..........  5% 4.11M 4s
   450K .......... .......... .......... .......... ..........  6% 3.13M 4s
   500K .......... .......... .......... .......... ..........  6% 62.2M 3s
   550K .......... .......... .......... .......... ..........  7% 2.58M 3s
   600K .......... .......... .......... .......... ..........  7% 95.8M 3s
   650K .......... .......... .......... .......... ..........  8% 12.6M 3s
   700K .......... .......... .......... .......... ..........  9% 44.5M 3s
   750K .......... .......... .......... .......... ..........  9%  107M 3s
   800K .......... .......... .......... .......... .......... 10% 2.97M 2s
   850K .......... .......... .......... .......... .......... 11% 47.3M 2s
   900K .......... .......... .......... .......... .......... 11% 3.73M 2s
   950K .......... .......... .......... .......... .......... 12% 40.6M 2s
  1000K .......... .......... .......... .......... .......... 12%  145M 2s
  1050K .......... .......... .......... .......... .......... 13%  108M 2s
  1100K .......... .......... .......... .......... .......... 14% 3.48M 2s
  1150K .......... .......... .......... .......... .......... 14% 36.8M 2s
  1200K .......... .......... .......... .......... .......... 15% 94.8M 2s
  1250K .......... .......... .......... .......... .......... 15%  123M 2s
  1300K .......... .......... .......... .......... .......... 16% 5.33M 2s
  1350K .......... .......... .......... .......... .......... 17% 76.5M 2s
  1400K .......... .......... .......... .......... .......... 17% 63.4M 2s
  1450K .......... .......... .......... .......... .......... 18% 99.3M 1s
  1500K .......... .......... .......... .......... .......... 19%  120M 1s
  1550K .......... .......... .......... .......... .......... 19% 3.45M 1s
  1600K .......... .......... .......... .......... .......... 20% 85.1M 1s
  1650K .......... .......... .......... .......... .......... 20% 83.7M 1s
  1700K .......... .......... .......... .......... .......... 21% 54.1M 1s
  1750K .......... .......... .......... .......... .......... 22%  153M 1s
  1800K .......... .......... .......... .......... .......... 22% 3.31M 1s
  1850K .......... .......... .......... .......... .......... 23% 96.4M 1s
  1900K .......... .......... .......... .......... .......... 23% 8.19M 1s
  1950K .......... .......... .......... .......... .......... 24% 3.93M 1s
  2000K .......... .......... .......... .......... .......... 25% 23.4M 1s
  2050K .......... .......... .......... .......... .......... 25% 75.4M 1s
  2100K .......... .......... .......... .......... .......... 26% 57.7M 1s
  2150K .......... .......... .......... .......... .......... 27% 79.2M 1s
  2200K .......... .......... .......... .......... .......... 27%  125M 1s
  2250K .......... .......... .......... .......... .......... 28% 74.9M 1s
  2300K .......... .......... .......... .......... .......... 28% 1.54M 1s
  2350K .......... .......... .......... .......... .......... 29% 76.6M 1s
  2400K .......... .......... .......... .......... .......... 30% 77.8M 1s
  2450K .......... .......... .......... .......... .......... 30%  125M 1s
  2500K .......... .......... .......... .......... .......... 31%  105M 1s
  2550K .......... .......... .......... .......... .......... 31%  152M 1s
  2600K .......... .......... .......... .......... .......... 32% 9.66M 1s
  2650K .......... .......... .......... .......... .......... 33% 60.8M 1s
  2700K .......... .......... .......... .......... .......... 33% 57.9M 1s
  2750K .......... .......... .......... .......... .......... 34% 26.0M 1s
  2800K .......... .......... .......... .......... .......... 35% 33.3M 1s
  2850K .......... .......... .......... .......... .......... 35% 28.8M 1s
  2900K .......... .......... .......... .......... .......... 36% 29.7M 1s
  2950K .......... .......... .......... .......... .......... 36% 20.7M 1s
  3000K .......... .......... .......... .......... .......... 37% 12.9M 1s
  3050K .......... .......... .......... .......... .......... 38% 3.07M 1s
  3100K .......... .......... .......... .......... .......... 38%  120M 1s
  3150K .......... .......... .......... .......... .......... 39% 8.25M 1s
  3200K .......... .......... .......... .......... .......... 39% 42.7M 1s
  3250K .......... .......... .......... .......... .......... 40% 76.2M 1s
  3300K .......... .......... .......... .......... .......... 41% 53.5M 1s
  3350K .......... .......... .......... .......... .......... 41% 22.3M 1s
  3400K .......... .......... .......... .......... .......... 42% 34.0M 1s
  3450K .......... .......... .......... .......... .......... 43% 59.8M 1s
  3500K .......... .......... .......... .......... .......... 43% 60.1M 1s
  3550K .......... .......... .......... .......... .......... 44% 53.6M 1s
  3600K .......... .......... .......... .......... .......... 44%  104M 1s
  3650K .......... .......... .......... .......... .......... 45% 69.5M 1s
  3700K .......... .......... .......... .......... .......... 46% 40.6M 1s
  3750K .......... .......... .......... .......... .......... 46%  141M 1s
  3800K .......... .......... .......... .......... .......... 47% 41.1M 1s
  3850K .......... .......... .......... .......... .......... 47%  153M 1s
  3900K .......... .......... .......... .......... .......... 48%  102M 1s
  3950K .......... .......... .......... .......... .......... 49% 28.8M 0s
  4000K .......... .......... .......... .......... .......... 49% 42.1M 0s
  4050K .......... .......... .......... .......... .......... 50% 8.85M 0s
  4100K .......... .......... .......... .......... .......... 51% 76.4M 0s
  4150K .......... .......... .......... .......... .......... 51% 42.6M 0s
  4200K .......... .......... .......... .......... .......... 52%  145M 0s
  4250K .......... .......... .......... .......... .......... 52% 40.7M 0s
  4300K .......... .......... .......... .......... .......... 53%  165M 0s
  4350K .......... .......... .......... .......... .......... 54% 33.2M 0s
  4400K .......... .......... .......... .......... .......... 54% 59.7M 0s
  4450K .......... .......... .......... .......... .......... 55% 22.3M 0s
  4500K .......... .......... .......... .......... .......... 55% 7.38M 0s
  4550K .......... .......... .......... .......... .......... 56% 49.4M 0s
  4600K .......... .......... .......... .......... .......... 57% 49.3M 0s
  4650K .......... .......... .......... .......... .......... 57% 23.7M 0s
  4700K .......... .......... .......... .......... .......... 58% 7.84M 0s
  4750K .......... .......... .......... .......... .......... 59% 2.03M 0s
  4800K .......... .......... .......... .......... .......... 59% 73.0M 0s
  4850K .......... .......... .......... .......... .......... 60% 52.2M 0s
  4900K .......... .......... .......... .......... .......... 60% 86.3M 0s
  4950K .......... .......... .......... .......... .......... 61% 86.9M 0s
  5000K .......... .......... .......... .......... .......... 62% 58.9M 0s
  5050K .......... .......... .......... .......... .......... 62% 45.5M 0s
  5100K .......... .......... .......... .......... .......... 63% 58.5M 0s
  5150K .......... .......... .......... .......... .......... 63% 13.6M 0s
  5200K .......... .......... .......... .......... .......... 64% 12.1M 0s
  5250K .......... .......... .......... .......... .......... 65% 8.82M 0s
  5300K .......... .......... .......... .......... .......... 65% 22.5M 0s
  5350K .......... .......... .......... .......... .......... 66% 3.66M 0s
  5400K .......... .......... .......... .......... .......... 67% 15.7M 0s
  5450K .......... .......... .......... .......... .......... 67% 22.0M 0s
  5500K .......... .......... .......... .......... .......... 68% 12.7M 0s
  5550K .......... .......... .......... .......... .......... 68% 32.0M 0s
  5600K .......... .......... .......... .......... .......... 69%  121M 0s
  5650K .......... .......... .......... .......... .......... 70% 19.4M 0s
  5700K .......... .......... .......... .......... .......... 70% 16.5M 0s
  5750K .......... .......... .......... .......... .......... 71%  126M 0s
  5800K .......... .......... .......... .......... .......... 71% 26.8M 0s
  5850K .......... .......... .......... .......... .......... 72% 15.1M 0s
  5900K .......... .......... .......... .......... .......... 73%  118M 0s
  5950K .......... .......... .......... .......... .......... 73% 16.2M 0s
  6000K .......... .......... .......... .......... .......... 74% 37.5M 0s
  6050K .......... .......... .......... .......... .......... 75% 38.7M 0s
  6100K .......... .......... .......... .......... .......... 75% 39.4M 0s
  6150K .......... .......... .......... .......... .......... 76%  107M 0s
  6200K .......... .......... .......... .......... .......... 76% 32.6M 0s
  6250K .......... .......... .......... .......... .......... 77% 28.9M 0s
  6300K .......... .......... .......... .......... .......... 78% 59.8M 0s
  6350K .......... .......... .......... .......... .......... 78% 27.1M 0s
  6400K .......... .......... .......... .......... .......... 79% 23.0M 0s
  6450K .......... .......... .......... .......... .......... 79% 27.9M 0s
  6500K .......... .......... .......... .......... .......... 80% 20.6M 0s
  6550K .......... .......... .......... .......... .......... 81%  142M 0s
  6600K .......... .......... .......... .......... .......... 81% 26.8M 0s
  6650K .......... .......... .......... .......... .......... 82% 99.7M 0s
  6700K .......... .......... .......... .......... .......... 83% 31.8M 0s
  6750K .......... .......... .......... .......... .......... 83% 21.3M 0s
  6800K .......... .......... .......... .......... .......... 84% 60.9M 0s
  6850K .......... .......... .......... .......... .......... 84% 29.0M 0s
  6900K .......... .......... .......... .......... .......... 85% 25.7M 0s
  6950K .......... .......... .......... .......... .......... 86% 5.46M 0s
  7000K .......... .......... .......... .......... .......... 86% 60.0M 0s
  7050K .......... .......... .......... .......... .......... 87% 99.8M 0s
  7100K .......... .......... .......... .......... .......... 87% 97.3M 0s
  7150K .......... .......... .......... .......... .......... 88%  135M 0s
  7200K .......... .......... .......... .......... .......... 89% 83.4M 0s
  7250K .......... .......... .......... .......... .......... 89% 48.4M 0s
  7300K .......... .......... .......... .......... .......... 90% 47.7M 0s
  7350K .......... .......... .......... .......... .......... 91% 68.5M 0s
  7400K .......... .......... .......... .......... .......... 91% 58.7M 0s
  7450K .......... .......... .......... .......... .......... 92%  132M 0s
  7500K .......... .......... .......... .......... .......... 92% 62.8M 0s
  7550K .......... .......... .......... .......... .......... 93% 10.9M 0s
  7600K .......... .......... .......... .......... .......... 94% 58.1M 0s
  7650K .......... .......... .......... .......... .......... 94%  169M 0s
  7700K .......... .......... .......... .......... .......... 95% 59.9M 0s
  7750K .......... .......... .......... .......... .......... 95%  134M 0s
  7800K .......... .......... .......... .......... .......... 96% 78.7M 0s
  7850K .......... .......... .......... .......... .......... 97% 57.8M 0s
  7900K .......... .......... .......... .......... .......... 97% 64.7M 0s
  7950K .......... .......... .......... .......... .......... 98% 15.2M 0s
  8000K .......... .......... .......... .......... .......... 99%  147M 0s
  8050K .......... .......... .......... .......... .......... 99% 37.6M 0s
  8100K .......... .......... .....                           100%  140M=0.7s

2016-02-22 19:23:19 (12.1 MB/s) - 'minecraft/minecraft_server.jar' saved [8320755/8320755]

 ---> 7cb305bef8de
Removing intermediate container 3e88b3ccc3f3
Step 7 : RUN echo "eula=true" > eula.txt
 ---> Running in 63762133b2f1
 ---> d7a07632cc01
Removing intermediate container 63762133b2f1
Step 8 : CMD java -Xms512m -Xmx1024m -jar minecraft/minecraft_server.jar nogui
 ---> Running in 365e5caac790
 ---> 03236eefff82
Removing intermediate container 365e5caac790
Step 9 : EXPOSE 25565
 ---> Running in b090537cd744
 ---> 20c4e9a517b6
Removing intermediate container b090537cd744
Successfully built 20c4e9a517b6
minecraft@minecraft-VirtualBox:~/minecraft-project/minecraft$ sudo docker run -i -t -p=25565:25565 minecraft/minecraft189
/bin/sh: 1: java: not found
`

**Comment by hayday100** [comment]
There is not color above but the errors in red are actually:

debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin:
Fetched 1040 kB in 0s (1893 kB/s)

**Comment by bilalakil** [comment]
As the error suggests, Java was not found. The Minecraft Server runs on Java.

From skimming through the log you provided I couldn't see any point in which you installed Java. Either install it yourself or change your base image from plain Ubuntu to one with Java installed, i.e.:

https://hub.docker.com/_/java/

**Comment by hayday100** [comment]
bilalakil: you got it precisely.  The kernel must be downgraded to 3.13.0-71.  There were a few hiccups while trying to do that.  First, the generic kernel downgraded only to 3.13.0-79 so I had to manually download and downgrade.  Then the virtualbox lost its mouse pointer.  I later found that you just have to use "input -> mouse integration" to reset the pointing device.  Thirdly, the docker daemon refused to start again after the downgrade and apt-get did not help.  I went to the docker website and followed the longer process to reinstall docker.  But finally everything works like a charm!  You don't need the workaround, just a good old engine!  Thanks to all!

**Comment by sjortiz** [comment]
This is still happening on Debian and the workaround doesn't fix it

**Comment by abdennour** [comment]
Build your image on top of it.

```
FROM java:openjdk-8-jdk
RUN sed -i "2 iexport JAVA_HOME=${JAVA_HOME}" /etc/ca-certificates/update.d/docker-openjdk
```
at 2nd line, it will insert JAVA_HOME with a computed value ( that's why double quotes not single quotes)


Credits to : https://github.com/jenkinsci/docker/issues/901
